storage/src/adb/keyless.rs

```diff
@@ -165,75 +160,184 @@ impl<E: Storage + Clock + Metrics, V: Codec, H: CHasher> Keyless<E, V, H> {
+                pin_mut!(stream);
+                stream.next().await
+            };
+            let Some(op) = op else {
+                warn!("no starting log operation found, returning empty db");
+                return Ok(Self {
+                    mmr,
+                    log,
+                    size: 0,
+                    locations,
+                    log_items_per_section,
+                    hasher,
+                    last_commit_loc: None,
+                });
+            };
+            let op = op?;
+            mmr.add_batched(&mut hasher, &op.3.encode()).await?;
+            mmr.sync(&mut hasher).await?;
+            locations.append(op.1).await?;
+            locations.sync().await?;
+
+            (op.0, op.1)
+        };
+
+        // Next, replay any log operations that are missing from the other structures, returning the
+        // operation at the log's tip.
+        let last_log_op = {
+            let stream = log.replay(section, offset, REPLAY_BUFFER_SIZE).await?;
+            pin_mut!(stream);
+            let op = stream.next().await;
+            // Very first operation is known to exist & be valid.
+            let op = op.expect("operation known to exist");
+            let mut op = op?.3;
+            while let Some(result) = stream.next().await {
+                let (section, offset, _, next_op) = result?;
+                let encoded_op = next_op.encode();
+                op = next_op;
                 warn!(
-                    old_size = locations_size,
-                    new_size = last_commit_loc + 1,
-                    "rewinding to last commit point"
+                    location = mmr.leaves(),
+                    section, offset, "adding missing operation to MMR/location map"
                 );
-                locations.rewind(last_commit_loc + 1).await?;
+                mmr.add_batched(&mut hasher, &encoded_op).await?;
+                locations.append(offset).await?;
+            }
+            if mmr.is_dirty() {
+                mmr.sync(&mut hasher).await?;
                 locations.sync().await?;
-                mmr.pop((locations_size - last_commit_loc - 1) as usize)
-                    .await?;
-                locations_size = last_commit_loc + 1;
-                // Rewind the operations log last to ensure the locations journal always references
-                // valid data in the event of failures.
-                let rewind_point = rewind_point.expect("no rewind point found");
-                let section = rewind_point.0 / cfg.log_items_per_section.get();
-                log.rewind_to_offset(section, rewind_point.1).await?;
-                log.sync(section).await?;
             }
-        } else if locations_size > 0 {
-            warn!(
-                old_size = locations_size,
-                "no commit point found, rewinding to start"
-            );
-            locations.rewind(0).await?;
+
+            op
+        };
+
```

Comment:
nit: remove `\n`

Reply:
removed & moved the op_index init after the below comment where it belongs.